km/km_hcalls.c

```diff
@@ -2085,6 +2086,121 @@ static km_hc_ret_t mlock_hcall(void* vcpu, int hc, km_hc_args_t* arg)
    return HC_CONTINUE;
 }
 
+/*
+ * Count of the number of active io contexts.
+ * Use this to prevent snapshots while io contexts are alive.
+ */
+unsigned int km_io_context_count = 0;
+
+/*
+ * long io_setup(unsigned int nr_events, aio_context_t *ctx_idp);
+ */
+static km_hc_ret_t io_setup_hcall(void* vcpu, int hc, km_hc_args_t* arg)
+{
+   void* ctx_idp = km_gva_to_kma(arg->arg2);
+   if (ctx_idp == NULL) {
+      arg->hc_ret = -EFAULT;
+   } else {
+      arg->hc_ret = __syscall_2(SYS_io_setup, arg->arg1, (uint64_t)ctx_idp);
+      if (arg->hc_ret == 0) {
+         __atomic_add_fetch(&km_io_context_count, 1, __ATOMIC_SEQ_CST);
+      }
+   }
+   return HC_CONTINUE;
+}
+
+/*
+ * int io_submit(aio_context_t ctx_id, long nr, struct iocb **iocbpp);
+ */
+static km_hc_ret_t io_submit_hcall(void* vcpu, int hc, km_hc_args_t* arg)
+{
+   struct iocb km_iocb_copy[arg->arg2];
+   struct iocb* km_iocb_list[arg->arg2];
+   struct iocb** local_iocbpp = km_gva_to_kma(arg->arg3);
+   if (local_iocbpp == NULL) {
+      arg->hc_ret = -EFAULT;
+      return HC_CONTINUE;
+   }
+
+   // xlate guest address in the iocb's to km address
+   for (int i = 0; i < arg->arg2; i++) {
+      km_iocb_copy[i] = *local_iocbpp[i];
+      km_iocb_list[i] = &km_iocb_copy[i];
+      km_iocb_copy[i].aio_buf = (uint64_t)km_gva_to_kma(local_iocbpp[i]->aio_buf);
+      if (km_iocb_copy[i].aio_buf == (uint64_t)NULL) {
+         arg->hc_ret = -EFAULT;
+         return HC_CONTINUE;
+      }
+   }
+   arg->hc_ret = __syscall_3(SYS_io_submit, arg->arg1, arg->arg2, (uint64_t)&km_iocb_list);
+
+   // Give back kernel supplied values.
+   for (int i = 0; i < arg->arg2; i++) {
+      local_iocbpp[i]->aio_key = km_iocb_copy[i].aio_key;
+   }
+
+   return HC_CONTINUE;
+}
+
+/*
+ * int syscall(SYS_io_cancel, aio_context_t ctx_id, struct iocb *iocb,
+ *             struct io_event *result);
+ */
+static km_hc_ret_t io_cancel_hcall(void* vcpu, int hc, km_hc_args_t* arg)
+{
+   struct iocb* local_iocbp = km_gva_to_kma(arg->arg2);
+   struct io_event* local_resultp = km_gva_to_kma(arg->arg3);
+   if (local_iocbp == NULL || local_resultp == NULL) {
+      arg->hc_ret = -EFAULT;
+      return HC_CONTINUE;
+   }
+   arg->hc_ret = __syscall_3(SYS_io_cancel, arg->arg1, (uint64_t)local_iocbp, (uint64_t)local_resultp);
+   return HC_CONTINUE;
+}
+
+/*
+ * int syscall(SYS_io_destroy, aio_context_t ctx_id);
+ */
+static km_hc_ret_t io_destroy_hcall(void* vcpu, int hc, km_hc_args_t* arg)
+{
+   arg->hc_ret = __syscall_1(SYS_io_destroy, arg->arg1);
+   if (arg->hc_ret == 0) {
+      km_assert(__atomic_load_n(&km_io_context_count, __ATOMIC_SEQ_CST) > 0);
```

Comment:
is there a race condition between assert and decrement?

Reply:
Yes.  Potentially one thread could be creating an i/o context while another is destroying one.
We can't trust the assertion here.

Reply:
I take my reply back.  We are only testing to ensure that the active context count doesn't go negative.
I think it should never go negative.